packages/jellyfish-api-core/__tests__/category/loan/updateVault.test.ts

```diff
@@ -0,0 +1,466 @@
+  })
+
+  it('should not updateVault if loanSchemeId is invalid', async () => {
+    const promise = alice.rpc.loan.updateVault(vaultId, {
+      loanSchemeId: 'INVALID_SCHEME_ID'
+    })
+
+    await expect(promise).rejects.toThrow('RpcApiError: \'Test UpdateVaultTx execution failed:\nCannot find existing loan scheme with id INVALID_SCHEME_ID\', code: -32600, method: updatevault')
+  })
+
+  it('should not updateVault if owner address and loanSchemeId are not set', async () => {
+    const promise = alice.rpc.loan.updateVault(vaultId, {})
+
+    await expect(promise).rejects.toThrow('RpcApiError: \'At least ownerAddress OR loanSchemeId must be set\', code: -8, method: updatevault')
+  })
+
+  it('should not updateVault if any of the asset\'s price is invalid', async () => {
+    // Create another vault
+    const vaultId = await alice.rpc.loan.createVault({
+      ownerAddress: address,
+      loanSchemeId: 'default'
+    })
+    await alice.generate(1)
+
+    // Deposit to vault
+    await alice.rpc.loan.depositToVault({
+      vaultId, from: address, amount: '100@DFI'
+    })
+    await alice.generate(1)
+
+    // Take loan
+    await alice.rpc.loan.takeLoan({
+      vaultId,
+      amounts: '50@TSLA'
+    })
+    await alice.generate(1)
+
+    await alice.rpc.oracle.setOracleData(oracleId, Math.floor(new Date().getTime() / 1000), { prices: [{ tokenAmount: '0.1@DFI', currency: 'USD' }] })
+
+    // Wait for the price become invalid
+    await alice.container.waitForPriceInvalid('DFI/USD')
+
+    // Frozen = invalid price and active vault
+    const data = await alice.rpc.loan.getVault(vaultId)
+    expect(data.state).toStrictEqual(VaultState.FROZEN)
+
+    {
+      // Unable to update to the new scheme as vault state is frozen
+      const promise = alice.rpc.loan.updateVault(vaultId, {
+        ownerAddress: await alice.generateAddress(),
+        loanSchemeId: 'default'
+      })
+      await expect(promise).rejects.toThrow('RpcApiError: \'Test UpdateVaultTx execution failed:\nCannot update vault while any of the asset\'s price is invalid\', code: -32600, method: updatevault')
+    }
+
+    await alice.container.waitForPriceValid('DFI/USD')
+
+    // Update back the price
+    await alice.rpc.oracle.setOracleData(oracleId, Math.floor(new Date().getTime() / 1000), { prices: [{ tokenAmount: '1@DFI', currency: 'USD' }] })
+    await alice.generate(1)
+
+    // Wait for the price becomes invalid again
+    await alice.container.waitForPriceInvalid('DFI/USD')
+
+    {
+      // Unable to update to a new scheme as price is invalid and vault is going to be liquidated
+      const promise = alice.rpc.loan.updateVault(vaultId, {
+        ownerAddress: await alice.generateAddress(),
+        loanSchemeId: 'default'
+      })
+      await expect(promise).rejects.toThrow('RpcApiError: \'Vault is under liquidation.\', code: -26, method: updatevault')
```

Comment:
The logic flow here confused me, L264 is waiting the price to be invalid, then error here became vault under liquidation. Liquidation itself should not happen when price invalid isn't it?

Since there is another test L312 test against liquidated vault, why this test is needed here? (and combined with FROZEN vault test)

Reply:
Update: Saw other's comment, if the purpose is to test `mayLiquidate` state, should happen while price is valid.

Reply:
1) The logic flow here confused me:  (Fixed. Ya I made a wrong assumption here.)
2) Added a test item to test `mayliquidate` => `should updateVault if the vault state is mayliquidate`